File: py/D2.py
from dataclasses import dataclass
import re

@dataclass
class policy:
    lo: int
    hi: int
    ch: str
    pw: str

def parseInput():
    file = open("../inputs/D2.input", "r")
    policies = []
    contents = file.read()
    for line in contents.splitlines():
        matchGroups = re.search("([0-9]+)-([0-9]+) ([a-z]): ([a-z]+)", line).groups()
        policies.append(policy(
            int(matchGroups[0]),
            int(matchGroups[1]),
            matchGroups[2],
            matchGroups[3]
        ))
    return policies

def p1():
    policies = parseInput()
    valid = 0
    for policy in policies:
        count = 0
        for char in policy.pw:
            if (char == policy.ch):
                count += 1
                if (count > policy.hi):
                    break
        
        if (count >= policy.lo and count <= policy.hi):
            valid += 1

    return valid
def p2():
    policies = parseInput()
    valid = 0
    for policy in policies:
        if ((policy.pw[policy.lo - 1] == policy.ch) 
         != (policy.pw[policy.hi - 1] == policy.ch)):
            valid += 1
    return valid

print(p1())
print(p2())

File: rs/src/d2.rs
use regex::Regex;
use std::fs;

struct Policy {
    lo: i32,
    hi: i32,
    ch: char,
    pw: String,
}

pub fn p1() -> i32 {
    let policies = parse_input();
    let mut valid = 0;

    for policy in policies {
        let mut count = 0;

        for ch in policy.pw.chars() {
            if ch == policy.ch {
                count += 1;
            }
        }

        if count >= policy.lo && count <= policy.hi {
            valid += 1;
        }
    }

    return valid;
}

pub fn p2() -> i32 {
    let policies = parse_input();
    let mut valid = 0;

    for policy in policies {
        let lo_match = policy.ch == policy.pw.chars().nth((policy.lo - 1) as usize).unwrap();
        let hi_match = policy.ch == policy.pw.chars().nth((policy.hi - 1) as usize).unwrap();
        if lo_match ^ hi_match {
            valid += 1;
        }
    }

    return valid;
}

fn parse_input() -> Vec<Policy> {
    let contents = fs::read_to_string("../inputs/D2.input").unwrap();
    
    let mut policies = Vec::new();
    let regex = Regex::new("^([0-9]+)-([0-9]+) ([a-z]): ([a-z]+)$").unwrap();
    for line in contents.split("\r\n") {
        let capture = regex.captures(line).unwrap();
        let lo = capture[1].parse::<i32>().unwrap();
        let hi = capture[2].parse::<i32>().unwrap();
        let ch = capture[3].chars().nth(0).unwrap();
        let pw = capture[4].to_string();
        policies.push(Policy
        {
            lo: lo,
            hi: hi,
            ch: ch,
            pw: pw
        });
    }

    return policies;
}